### PR TITLE
Attempting to fix publishing

### LINKED ISSUE #9129: ci: `unleash-check` is failing in master pipeline
At the time of writing, the `unleash-check` job has been failing probably since https://github.com/paritytech/substrate/pull/8472 ([CI job](https://gitlab.parity.io/parity/substrate/-/jobs/875772)).

The immediate reason for that is enabling a [new Cargo feature resolver](https://blog.rust-lang.org/2021/03/25/Rust-1.51.0.html#cargos-new-feature-resolver) shipped in Rust 1.51, while `cargo-unleash` couldn't parse the updated manifest file at that time due to not supporting new enough Cargo-the-library.

However, simply using an updated `cargo-unleash` (#8848) does not fix the problem on its own, due to other local changes made to the workspace in the meantime, while the CI job was red.

The check originally meant to verify that the current workspace is publishable and builds in isolation. However, in the meantime we introduced some backwards incompatible changes and it looks like this can have a cascade impact on the entire workspace but the full extent of required changes is not yet determined.

It's worth keeping in mind that in the future an innocuous change might have a big impact across the graph possibly requiring us to bump major version. This ties into the discussion on how we should manage crate versioning (e.g. #7634) in that case but the relevant discussion should probably be had elsewhere.

cc @vas3k @grbIzl @gnunicorn

### PR DESCRIPTION
**Pull request by gnunicorn**
This sort of works locally, let's see what the CI thinks about that.

Fixes #9129 
//cc @xanewok

### CODE REVIEW

**Comment by Xanewok** [comment]
The `unleash-check` is [green](https://gitlab.parity.io/parity/substrate/-/jobs/974907) :tada: 

This takes a crude approach of detecting which packages have changed since `v3.0.0` tag and applies a bump to a next breaking version (for version `A.B.C` bumps A if A > 0 and bumps B otherwise). If a package has not been released yet, e.g. `max-encoded-len-derive`, we don't need to od the breaking bump and instead we still have freedom to update the contents here without changing the version and before finally releasing it.

It's worth noting that this does not bump every crate to a unified `4.0.0-dev` or `0.10.0-dev` where applicable but only bumps where possible for the check to success (some versions are unchanged).
More importantly, this can sometimes lead to duplicated `{sc,sp}-*` crates when checking single packages with unleash. Our workspace view of the packages does not contain duplicates in the lockfile but I think we need to double-check that we don't introduce any duplicates when pulling these packages once they are uploaded.

**Comment by bkchr** [comment]
> max-encoded-len-derive

This should not be published!

**Comment by Xanewok** [comment]
> > max-encoded-len-derive
> 
> This should not be published!

I'm seeing that it's used (via `derive` feature) by crates that are meant to be published like [`pallet-assets`](https://github.com/paritytech/substrate/blob/d03a91a181d0b22d00a6b9ba2a8007dc254779e3/frame/assets/Cargo.toml#L25)or [`sp-core`](https://github.com/paritytech/substrate/blob/d03a91a181d0b22d00a6b9ba2a8007dc254779e3/primitives/core/Cargo.toml#L43) to name a few.  Don't we need to publish this the `-derive` crate as well? I remember that it wasn't possible to publish a proc macro crate included as part of another one in crates.io.

**Comment by bkchr** [comment]
The idea is that this functionality will be moved to scale-codec, but yeah not sure when this will happen... So yeah, we will probably need to release this..

**Comment by coriolinus** [comment]
https://github.com/paritytech/parity-scale-codec/pull/268 is still open and shouldn't take a ton of effort to update, if we do wish to move everything `MaxEncodedLen` into `parity-scale-codec`.

**Comment by Xanewok** [comment]
#9163 is finally merged and the `max-encoded-len` is migrated into `parity-scale-codec` so let's see what CI has still to say.

**Comment by Xanewok** [comment]
Right now `gitlab-check-polkadot-companion-build ` fails but that's because we pull duplicate dependencies via BEEFY (grandpa-bridge-gadget) crates. Once these are patched to a local checkout everything compiles.

**Comment by Xanewok** [comment]
This is now ready for review.

https://github.com/paritytech/substrate/pull/9140#issuecomment-875062964 is addressed by explicitly running `cargo update -p sp-core` because otherwise the patching done by `diener` in the Polkadot CI job will not be taken into account, leading to duplicated dependencies (and thus failing the build).

It'd be good to land that sooner rather than later. Firstly, it's prone to bitrotting as any change to a Cargo.toml in the meantime probably requires this PR to be updated. Secondly, it'd be good to finally get the CI green for master and to mark the `unleash-check` job as required again (cc @paritytech/ci)

Took this over from @gnunicorn so I can approve it on his behalf if it's well-reviewed :smiley:

File: .gitlab-ci.yml
Comment (ordian) [inline .gitlab-ci.yml:335]: why not put in the `variables:` section above?

File: .gitlab-ci.yml
Comment (ordian) [inline .gitlab-ci.yml:43]: it's ironic the tool aimed to fix versioning is unversioned

Comment (Xanewok) [reply .gitlab-ci.yml:43]: We agreed with @gnunicorn to release a new version once we ensure that it's capable of fixing this versioning issue here.

Should we do that and switch to the new version, then?

File: bin/node-template/pallets/template/Cargo.toml
Comment (ordian) [inline bin/node-template/pallets/template/Cargo.toml:24]: does it make sense to omit versions for dev-deps only leaving path as you did in https://github.com/paritytech/parity-common/pull/552#discussion_r661848985?

File: .gitlab-ci.yml
Comment (Xanewok) [reply .gitlab-ci.yml:335]: That's a makeshift optimization that needs addressing in the tool itself and didn't want to "properly" encode it. Should I move it above for consistency for now?

File: bin/node-template/pallets/template/Cargo.toml
Comment (Xanewok) [reply bin/node-template/pallets/template/Cargo.toml:24]: That's something that I have discovered recently and will probably impact the unleash tool as well. As of now, `cargo unleash` automatically removes dev-dependencies to break these cycles but if there is 1st party support for that in Cargo than we should rethink what `cargo unleash` does and probably migrate to `path`-only dev-dependencies in the Substrate repository in the near future.

We talked with @gnunicorn already about that and we will address that somehow but for now I would prefer not to touch that in this PR.

File: .gitlab-ci.yml
Comment (ordian) [reply .gitlab-ci.yml:335]: I see, it's fine to leave it as is then with the comment it's going to removed once cargo unleash addresses the issue.

**Review by ordian — APPROVED** [review]

File: .gitlab-ci.yml
Comment (Xanewok) [reply .gitlab-ci.yml:336]: Added a note

File: .gitlab-ci.yml
Comment (Xanewok) [reply .gitlab-ci.yml]: There was a release of `.alpha.12` in the meantime that's upgraded to the new Cargo so I switched to use that instead for now

**Comment by Xanewok** [comment]
Merged master and the checks are green again; it'd be great to get one more approval and let's try merging it then

**Review by sorpaas — APPROVED** [review]

**Comment by Xanewok** [comment]
bot merge

**Comment by sorpaas** [comment]
Cheers!

**Pull request merged**
